src/main/java/org/spongepowered/api/text/template/TextTemplate.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    private void appendElement(Text.Builder builder, Object element, Map<String, Object> params) {
+        if (element instanceof TextElement) {
+            ((TextElement) element).applyTo(builder);
+        } else if (element instanceof Arg) {
+            Arg<?> arg = (Arg<?>) element;
+            Object param = params.get(arg.name);
+            if (param == null) {
+                if (!arg.optional) {
+                    throw new RequiredTextTemplateArgumentException(arg.name);
+                }
+            } else {
+                arg.eval(param).applyTo(builder);
+            }
+        } else {
+            builder.append(Text.of(element.toString()));
+        }
+    }
+
+    /**
+     * Constructs a new TextTemplate for the given elements. The order of the
+     * elements is the order in which they will be appended to the result
+     * builder via {@link #eval(Map)}.
+     *
+     * <p>The provided elements may be of any type.</p>
+     *
+     * <p>In the case that an element is a {@link TextElement},
+     * {@link TextElement#applyTo(Text.Builder)} will be used to append the
+     * element to the builder.</p>
+     *
+     * <p>In the case that an element is an {@link Arg} the argument will be
+     * replaced with a {@link TextElement} object via the {@link Function}
+     * provided by {@link #arg(String, Class, Function, boolean)}. If no
+     * function was provided the argument expects a TextElement type. The
+     * parameter value in the provided map in {@link #eval(Map)} must match the
+     * type of the input for the arguments Function or a
+     * {@link MismatchedTextTemplateArgumentException} will be thrown.</p>
+     *
+     * <p>In the case that an element is any other type, the parameter value's
+     * {@link Object#toString()} method will be used to create a {@link Text}
+     * object.</p>
+     *
+     * @param elements Elements to append to builder
+     * @return Newly constructed TextTemplate
+     */
+    public static TextTemplate of(Object... elements) {
+        checkNotNull(elements, "elements");
+        if (elements.length == 0) {
+            return EMPTY;
+        }
+        return new TextTemplate(elements);
+    }
+
+    /**
+     * Constructs a new {@link Arg} to be supplied to {@link #of(Object...)}.
+     * The argument evaluates it's output by applying a {@link Function} that
+     * takes any type as it's input and returns a {@link TextElement} as
+     * it's output.
+     *
+     * @param name Name of parameter
+     * @param type Type of parameter
+     * @param f Function that takes "type" as the input and returns Text as the output
+     * @param optional Whether this argument is allowed to be omitted
+     * @param <T> Input type of argument
+     * @return Newly constructed Arg
+     */
+    public static <T> Arg<T> arg(String name, Class<T> type, Function<T, TextElement> f, boolean optional) {
+        checkNotNull(name, "name");
```

Comment:
Yes, it safes you from creating a new instance in the case that some params are wrong. But invalid input is not to be expected in this method in prod systems. And on dev systems it does not matter since the bad call is a programming bug that must be fixed anyway and 10us are neglectible. And this prevents bypassed checks because something is added later on/called internally.